File: java-reactive/src/main/java/com/github/fidgetting/util/R2DBC.java
package com.github.fidgetting.util;

import io.r2dbc.pool.ConnectionPool;
import io.r2dbc.pool.ConnectionPoolConfiguration;
import io.r2dbc.postgresql.PostgresqlConnectionConfiguration;
import io.r2dbc.postgresql.PostgresqlConnectionFactory;

import java.time.Duration;

public class R2DBC {

  public static Pool createPool(String host, int port, String username, String password) {
    return new Pool(
        new ConnectionPool(
            ConnectionPoolConfiguration.builder(
                new PostgresqlConnectionFactory(PostgresqlConnectionConfiguration.builder()
                    .host(host)
                    .port(port)
                    .username(username)
                    .password(password)
                    .database("postgres")
                    .build()))
                .maxIdleTime(Duration.ofSeconds(1))
                .maxSize(10)
                .build()
        )
    );
  }

  public static class Pool implements AutoCloseable {

    public final ConnectionPool pool;

    public Pool(ConnectionPool pool) {
      this.pool = pool;
    }

    public void close() {
      pool.dispose();
    }
  }

}

File: scala/src/main/scala/com/github/fidgetting/util/R2DBC.scala
package com.github.fidgetting.util

import io.r2dbc.pool.{ConnectionPool, ConnectionPoolConfiguration}
import io.r2dbc.postgresql.{PostgresqlConnectionConfiguration, PostgresqlConnectionFactory}

import scala.concurrent.duration.{DurationInt, FiniteDuration}
import scala.jdk.javaapi.DurationConverters.toJava
import scala.util.Using.Releasable

object R2DBC {

  def createPool(
    host: String,
    port: Int,
    username: String,
    password: String,
    database: String = "postgres",
    idleTimeout: FiniteDuration = 1.second,
    maxSize: Int = 10,
  ): ConnectionPool = {
    ConnectionPool(
      ConnectionPoolConfiguration.builder(
          PostgresqlConnectionFactory(PostgresqlConnectionConfiguration.builder
            .host(host)
            .port(port)
            .username(username)
            .password(password)
            .database(database)
            .build()))
        .maxIdleTime(toJava(idleTimeout))
        .maxSize(maxSize)
        .build()
    )
  }

  given Releasable[ConnectionPool] with {
    def release(resource: ConnectionPool): Unit =
      resource.dispose()
  }

}
